### PR TITLE
feat(renderer): enhance conditions

### LINKED ISSUE #427: Allow Conditions to set form values
**Scope**:  FormRender, Conditions

**Description**

User may often want to use a condition to not only (optionally) evaluate the display criteria for a form field, but to also (optionally) set a form value. For example, if there is a simple 3 field form - field A is a number, field B is a string and field C should have its value set based on a set of conditions on both A and B.

This feature should accomodate the ability to

- evaluate a condition and set a form field without changing the visibility of the  field
- evaluate a condition and set a form field and change the visibility of the field
- support hidden, but enabled fields

**Semantics / Syntax Suggestion**

**<u>Evaluate a condition and set a form field without changing the visibility of the field</u>**

```javascript
{
  fields: [
    {
      name: 'Foo', // controlled field
      component: 'text-field',
      condition: {   // <-- new keyword 
        when: 'BarFoo', // name of controlled field
        is: 'Bar',      // condition
        remainVisible: true // remain visible (default)
        then: [{ // <-- absense of the "set" keyword means to set this field
          to: 'I have been set because BarFoo = Bar' // value to set
        },
        {
	 set: 'Woo',                                // name of field to set
          to: 'I have been set because BarFoo = Bar' // value to set
        }]
      }
    }, 
    {
      name: 'Woo', // controlled field
      component: 'text-field',
    },
    {
      name: 'BarFoo',
      label: 'Foo is Bar!',
      component: 'text-field',
    }
  ]
}
```

**<u>Evaluate a condition and set a form field and change the visibility of the field</u>**

```javascript
{
  fields: [
    {
      name: 'Foo', // controlled field
      component: 'text-field',
      condition: {   // <-- new keyword 
        when: 'BarFoo', // name of controlled field
        is: 'Bar',      // condition
        remainVisible: false // become visible when condition is met and set fields
        then: [{ // <-- absense of the "set" keyword means to set this field
          to: 'I have been set because BarFoo = Bar' // value to set
        }]
      }
    }, 
    {
      name: 'Woo', // controlled field
      component: 'text-field',
    },
    {
      name: 'BarFoo',
      label: 'Foo is Bar!',
      component: 'text-field',
    }
  ]
}
```

### PR DESCRIPTION
**Pull request by rvsia**
- adds sequence condition
- adds then, else branches
- adds set action
- optimize condition to listen values changes

close #427 

TODO:
- [x] tests
- [x] schema validator update
- [x] documentation
- [x] first render issue

### CODE REVIEW

File: packages/react-form-renderer/src/form-renderer/condition.js [hunk 1]
```diff
@@ -27,48 +29,130 @@ const fieldCondition = (value, { is, isNotEmpty, isEmpty, pattern, notMatch, fla
...
+            });
+          });
+        }
+      });
+      setSets(setters);
+    }
+  }, [setters, dirty]);
+
+  return conditionResult.visible ? children : null;
+}, isEqual);
```
Comment (Hyperkid123) [inline packages/react-form-renderer/src/form-renderer/condition.js:121]: I think this memo comparator is not going to work very well. Using just the `values` slightly more efficient. This way you compare all the children nodes of condition.

Comment (Hyperkid123) [reply packages/react-form-renderer/src/form-renderer/condition.js:121]: It should mainly prevent re-rendering the FormSpy wrapper which is the most expensive component.

File: packages/react-form-renderer/src/form-renderer/condition.js [hunk 2]
```diff
@@ -27,48 +29,137 @@ const fieldCondition = (value, { is, isNotEmpty, isEmpty, pattern, notMatch, fla
...
+            setTimeout(() => {
+              formOptions.batch(() => {
+                Object.entries(setter).forEach(([name, value]) => {
+                  formOptions.change(name, value);
+                });
+              });
+            });
+          }
+        });
+        setSets(setters);
```
Comment (Hyperkid123) [inline packages/react-form-renderer/src/form-renderer/condition.js:120]: we are doing two setters in one effect, that will cause additional render. Can we move from `useState` to `useReducer`?

File: package.json [hunk 3]
```diff
@@ -37,6 +37,7 @@
     }
   },
   "devDependencies": {
+    "@babel/plugin-transform-runtime": "^7.9.0",
```
Comment (Hyperkid123) [inline package.json:40]: is this new plugin necessary? I think that `babel-env` should provide same features.

**Comment by rvsia** [comment]
@Hyperkid123 

`useState` transformed to `useReducer`

babel-env is getting targets to enable async/await generators

**Comment by Hyperkid123** [comment]
> @Hyperkid123
> 
> `useState` transformed to `useReducer`
> 
> babel-env is getting targets to enable async/await generators

Riight that's it. :thinking: So it affects SSR and node dependency. Hm, I am thinking that we should rather use the plugin then. If might become an issue if somebody tries to deploy this on SSR env with an outdated version of node.

**Comment by rvsia** [comment]
@Hyperkid123 like next.js on firebase, right? :D 

Undropped!

**Pull request merged**